ambersim/trajopt/shooting.py

```diff
@@ -0,0 +1,170 @@
+from functools import partial
+from typing import Tuple
+
+import jax
+import jax.numpy as jnp
+from flax import struct
+from jax import lax, vmap
+from mujoco import mjx
+from mujoco.mjx import step
+
+from ambersim.trajopt.base import TrajectoryOptimizer, TrajectoryOptimizerParams
+from ambersim.trajopt.cost import CostFunction
+
+"""Shooting methods and their derived subclasses."""
+
+
+# ##### #
+# UTILS #
+# ##### #
+
+
+def shoot(m: mjx.Model, q0: jax.Array, v0: jax.Array, us: jax.Array) -> Tuple[jax.Array, jax.Array]:
+    """Utility function that shoots a model forward given a sequence of control inputs.
+
+    Args:
+        m: The model.
+        q0: The initial generalized coordinates.
+        v0: The initial generalized velocities.
+        us: The control inputs.
+
+    Returns:
+        qs (shape=(N + 1, nq)): The generalized coordinates.
+        vs (shape=(N + 1, nv)): The generalized velocities.
+    """
+    d = mjx.make_data(m)
```

Comment:
I make a new `Data` for every call to the `shoot` function (and in a `vmap`, the fields of `d` should become `vmap`ped as well, I believe). Is it really slow to do this? If so, how else should I be managing `mjx.Data`?

Reply:
Tested: doesn't make a difference. The main difference is in the `step` call.